File: 6_tree/LCA问题/tarjan离线/OfflineLCA.go
// https://ei1333.github.io/library/graph/tree/offline-lca.hpp
// O(n*α(n))，LCA离线，一般用于树上莫队

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	// https://judge.yosupo.jp/problem/lca
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	parents := make([]int, n-1)
	for i := 0; i < n-1; i++ {
		fmt.Fscan(in, &parents[i])
	}

	tree := make([][]int, n)
	for i := 0; i < n-1; i++ {
		tree[parents[i]] = append(tree[parents[i]], i+1)
		tree[i+1] = append(tree[i+1], parents[i])
	}

	queries := make([][2]int, q)
	for i := 0; i < q; i++ {
		var u, v int
		fmt.Fscan(in, &u, &v)
		queries[i] = [2]int{u, v}
	}

	res := OfflineLCA(tree, queries, 0)
	for i := 0; i < q; i++ {
		fmt.Fprintln(out, res[i])
	}
}

// LCA离线.
func OfflineLCA(tree [][]int, queries [][2]int, root int) []int {
	n := len(tree)
	ufa := NewUnionFindArray(n)
	st, mark, ptr, res := make([]int, n), make([]int, n), make([]int, n), make([]int, len(queries))
	for i := 0; i < len(queries); i++ {
		res[i] = -1
	}
	top := 0
	st[top] = root
	for _, q := range queries {
		mark[q[0]]++
		mark[q[1]]++
	}
	q := make([][][2]int, n)
	for i := 0; i < n; i++ {
		q[i] = make([][2]int, 0, mark[i])
		mark[i] = -1
		ptr[i] = len(tree[i])
	}
	for i := range queries {
		u, v := queries[i][0], queries[i][1]
		q[u] = append(q[u], [2]int{v, i})
		q[v] = append(q[v], [2]int{u, i})
	}
	run := func(u int) bool {
		for ptr[u] != 0 {
			v := tree[u][ptr[u]-1]
			ptr[u]--
			if mark[v] == -1 {
				top++
				st[top] = v
				return true
			}
		}
		return false
	}

	for top != -1 {
		u := st[top]
		if mark[u] == -1 {
			mark[u] = u
		} else {
			ufa.Union(u, tree[u][ptr[u]])
			mark[ufa.Find(u)] = u
		}

		if !run(u) {
			for _, v := range q[u] {
				if mark[v[0]] != -1 && res[v[1]] == -1 {
					res[v[1]] = mark[ufa.Find(v[0])]
				}
			}
			top--
		}
	}

	return res
}

type UnionFindArray struct {
	data []int
}

func NewUnionFindArray(n int) *UnionFindArray {
	data := make([]int, n)
	for i := 0; i < n; i++ {
		data[i] = -1
	}
	return &UnionFindArray{data: data}
}

func (ufa *UnionFindArray) Union(key1, key2 int) bool {
	root1, root2 := ufa.Find(key1), ufa.Find(key2)
	if root1 == root2 {
		return false
	}
	if ufa.data[root1] > ufa.data[root2] {
		root1, root2 = root2, root1
	}
	ufa.data[root1] += ufa.data[root2]
	ufa.data[root2] = root1
	return true
}

func (ufa *UnionFindArray) Find(key int) int {
	if ufa.data[key] < 0 {
		return key
	}
	ufa.data[key] = ufa.Find(ufa.data[key])
	return ufa.data[key]
}

File: 6_tree/LCA问题/tarjan离线/OfflineLCA.ts
// https://ei1333.github.io/library/graph/tree/offline-lca.hpp
// O(n*α(n))，LCA离线，一般用于树上莫队

/**
 * 离线求LCA.
 */
function offLineLca(
  tree: number[][],
  queries: [start: number, end: number][] | number[][],
  root = 0
): number[] {
  const n = tree.length
  const data = new Int32Array(n)
  for (let i = 0; i < n; ++i) data[i] = -1
  const stack = new Uint32Array(n)
  const mark = new Int32Array(n)
  const ptr = new Uint32Array(n)
  const res = Array(queries.length)
  for (let i = 0; i < queries.length; ++i) res[i] = -1

  let top = 0
  stack[top] = root
  queries.forEach(([start, end]) => {
    mark[start]++
    mark[end]++
  })

  const q: [next: number, ei: number][][] = Array(n)
  for (let i = 0; i < n; ++i) {
    q[i] = []
    mark[i] = -1
    ptr[i] = tree[i].length
  }
  queries.forEach(([start, end], i) => {
    q[start].push([end, i])
    q[end].push([start, i])
  })

  while (top !== -1) {
    const u = stack[top]
    const cache = tree[u]
    if (mark[u] === -1) {
      mark[u] = u
    } else {
      union(u, cache[ptr[u]])
      mark[find(u)] = u
    }

    if (!run(u)) {
      q[u].forEach(([v, i]) => {
        if (mark[v] !== -1 && res[i] === -1) {
          res[i] = mark[find(v)]
        }
      })
      --top
    }
  }

  return res

  function union(key1: number, key2: number): boolean {
    let root1 = find(key1)
    let root2 = find(key2)
    if (root1 === root2) return false
    if (data[root1] > data[root2]) {
      root1 ^= root2
      root2 ^= root1
      root1 ^= root2
    }
    data[root1] += data[root2]
    data[root2] = root1
    return true
  }

  function find(key: number): number {
    if (data[key] < 0) return key
    data[key] = find(data[key])
    return data[key]
  }

  function run(u: number): boolean {
    const cache = tree[u]
    while (ptr[u]) {
      const v = cache[--ptr[u]]
      if (mark[v] === -1) {
        stack[++top] = v
        return true
      }
    }
    return false
  }
}

export { offLineLca }

if (require.main === module) {
  const n = 5
  const edges = [
    [0, 1],
    [0, 2],
    [0, 3],
    [2, 4]
  ]
  const tree: number[][] = Array(n)
  for (let i = 0; i < n; ++i) tree[i] = []
  edges.forEach(([u, v]) => {
    tree[u].push(v)
    tree[v].push(u)
  })

  const queries = [
    [1, 3],
    [2, 4],
    [0, 4]
  ]
  const res = offLineLca(tree, queries)
  console.log(res)
}
